### PR TITLE
allow nonNullDefaults config

### PR DESCRIPTION
**Pull request by bastiankistner**
allow setting `nonNullDefaults`

### CODE REVIEW

**Comment by passionkind** [comment]
@JCMais could you please have a look at this PR and let me know if you'd require any changes ?

File: plugins/relay-node-interface/src/index.ts [hunk 1]
```diff
@@ -51,7 +58,7 @@ export function relayNodeInterfacePlugin(pluginConfig: RelayNodeInterfacePluginC
             name: 'Node',
             description: 'An object with a global ID',
             definition: (t) => {
-              t.id('id', {
+              ;(nonNullDefaults?.output ? t.nonNull : t.nullable).id('id', {
```
Comment (JCMais) [inline plugins/relay-node-interface/src/index.ts:61]: Given the default value above, `nonNullDefaults` will never be `null` / `undefined`, so the optional chaining is not needed.

File: plugins/relay-node-interface/src/index.ts [hunk 2]
```diff
@@ -70,7 +77,7 @@ export function relayNodeInterfacePlugin(pluginConfig: RelayNodeInterfacePluginC
             t.nullable.field('node', {
               type: 'Node',
               args: {
-                id: nonNull(
+                id: (nonNullDefaults?.input ? nonNull : (v: any) => v)(
```
Comment (JCMais) [inline plugins/relay-node-interface/src/index.ts:80]: same note than above.

File: plugins/relay-node-interface/src/index.ts [hunk 3]
```diff
@@ -1,6 +1,7 @@
 import { plugin, core, interfaceType, queryField, idArg, list, nonNull } from 'nexus'
 import { fromGlobalId } from 'graphql-relay'
 import { GraphQLResolveInfo } from 'graphql'
+import { NonNullConfig } from 'nexus/dist/core'
```
Comment (JCMais) [inline plugins/relay-node-interface/src/index.ts:4]: import just `core`, and then use `core.NonNullConfig`.

File: plugins/relay-global-id/src/index.ts [hunk 4]
```diff
@@ -1,6 +1,7 @@
 import { plugin, dynamicOutputMethod, core } from 'nexus'
 
 import { toGlobalId } from 'graphql-relay'
+import { NonNullConfig } from 'nexus/dist/core'
```
Comment (JCMais) [inline plugins/relay-global-id/src/index.ts:4]: import just `core`, and then use `core.NonNullConfig`

**Comment by passionkind** [comment]
@JCMais thank you for your constructive feedback! Applied changes accordingly.

One question: I figured that, when using the `globalId` plugin, outputs types are properly set. But root types are not being added. Is this intended or is this because `nexus-graphql` does not yet support dynamically adding root types? Or am I misunderstanding something?

Simple example:
- apply the `globalId` field somewhere => won't change the ts root type
- apply an `id` field somewhere => will add that field to the root types

**Comment by JCMais** [comment]
could you give me a concrete example of what you mentioned @passionkind?

the `globalId` field internally adds the `id` field (two if you are using `shouldAddRawId`)

**Comment by passionkind** [comment]
@JCMais please take a look at the following screenshot:

image

There you can see that 

`t.id('anotherId');` add the record key to `AccountUser` generated under `interface NexusGenObjects` 

However, `t.relayGlobalId('id');` does not add an `id` or even the `rawId` to the type.

But if we look at the field types interface `interface NexusGenFieldTypes`, there are both, the `anotherId`, the `id` and the `rawId`. As said, maybe I'm misunderstanding something. 
image

The resolver also does not complain when I define the `relayGlobalId` field as `nonNull`. But it does complain about the missing `anotherId`, when this is not being returned from the resolver.

In the following screenshot, I set bot explicitly as `nonNull` and you see the error message only mentioning the `anotherId`

image

Shouldn't there also be an error that the source type does not contain an `id` field, which would be turned into a base64 id by the plugin?

**Comment by JCMais** [comment]
Oh, that makes sense.

This is a decision by the Nexus team, please see this comment: https://github.com/graphql-nexus/nexus/issues/36#issuecomment-463649216

**Review by JCMais — APPROVED** [review]

**Comment by JCMais** [comment]
Thanks for the pull request @passionkind!

 I will get a release out this Friday.

**Comment by JCMais** [comment]
Looks like this was in fact a breaking change, fixed here: https://github.com/JCMais/nexus-plugins/commit/6b29ef948cf2892d099f5e514f8b6c3d83baf61e

I have enabled the workflows to run in PR to make it easier to see that in the future.

**Pull request merged**